src/main/java/seedu/nova/model/schedule/Schedule.java

```diff
@@ -0,0 +1,151 @@
+
+    private static final int DAYS_IN_WEEK = 7;
+    private static final int ACTUAL_RECESS_WEEK = 7;
+    private static final int SCHEDULE_RECESS_WEEK = 16;
+
+    private final LocalDate startDate;
+    private final LocalDate endDate;
+
+    private TreeMap<LocalDate, Week> mapMondayDateToWeek;
+
+    /**
+     * Instantiates a new Schedule.
+     *
+     * @param startDate the start date
+     * @param endDate   the end date
+     */
+    public Schedule(LocalDate startDate, LocalDate endDate) {
+
+        this.startDate = TimeUtil.getMondayOfWeek(startDate);
+        this.endDate = TimeUtil.getMondayOfWeek(endDate);
+
+        mapMondayDateToWeek = new TreeMap<>();
+    }
+
+    private Week getWeek(LocalDate sameWeekAs) {
+        LocalDate mondayOfSameWeek = TimeUtil.getMondayOfWeek(sameWeekAs);
+        return mapMondayDateToWeek.get(mondayOfSameWeek);
+    }
+
+    public Day getDay(LocalDate date) {
+        Week wk = getWeek(date);
+        if (wk == null) {
+            return null;
+        } else {
+            return wk.getDay(date.getDayOfWeek());
+        }
+    }
+
+    private Week getAndAddWeek(LocalDate date) {
+        Week wk = getWeek(date);
+        if (wk == null) {
+            wk = new Week(date);
+            mapMondayDateToWeek.put(TimeUtil.getMondayOfWeek(date), wk);
+        }
+        return wk;
+    }
+
+    /**
+     * Adds event.
+     *
+     * @param event the event
+     * @throws SchedulerException the command exception
+     */
+    public boolean addEvent(Event event) throws SchedulerException {
+        LocalDate date = event.getDate();
+        if (date.compareTo(startDate) < 0 || date.compareTo(endDate) > 0) {
+            throw new SchedulerException("Event date exceeded range");
+        }
+        return getAndAddWeek(date).addEvent(event);
+    }
+
+    /**
+     * Delete event.
+     *
+     * @param event the event
+     */
+    public boolean deleteEvent(Event event) {
+        LocalDate date = event.getDate();
+        Week wk = getWeek(date);
+        if (wk == null) {
+            return false;
+        } else {
+            return wk.deleteEvent(event);
+        }
+    }
+
+    /**
+     * Add lesson.
+     *
+     * @param lesson the lesson
+     * @throws SchedulerException the scheduler exception
+     */
+    public boolean addLesson(Lesson lesson) throws SchedulerException {
+        boolean ans = true;
+        for (int i = 0; i < 14; i++) {
+
+            if (i == ACTUAL_RECESS_WEEK) {
+                //No lesson on recess week
+                continue;
+            }
+
+            ans &= getAndAddWeek(startDate.plusWeeks(i)).addEvent(lesson);
+        }
+        return ans;
+    }
+
+    /**
+     * Views schedule on specified date.
+     *
+     * @param date the date
+     * @return the string
+     */
+    public String view(LocalDate date) {
+
+        Day d = getDay(date);
+        return Objects.requireNonNullElseGet(d, () -> new Day(date)).view();
```

Comment:
I'm not accepting this statement. There's really no need to do code optimization and now the code looks less reader friendly.